### PR TITLE
feat(internal): environment config for span sampling

### PR DESCRIPTION
**Pull request by ZStriker19**
## Description
In order for users to configure the single span sampling rules, this introduces the ability to do that via the environment variable `DD_SPAN_SAMPLING_RULES`. This PR also initializes the single span sampling processor if rules are provided.

## Motivation
This is required in order for users to specify the rules they would like to be applied for single span sampling.

## Design 
Most of the work introduced in this PR is done in sampling.py which is essentially responsible for: 
1. Pulling in and validating the rules from the environment `DD_SPAN_SAMPLING_RULES`
2. Providing those rules to the span processor so it can apply them to spans

## Testing strategy
Testing mostly revolves around making sure that different configurations are picked up correctly, default values are what we expect, and validation works as intended. 

In addition some tests were done around making sure that spans are actually sampled as expected, however a great deal of that has already been done in https://github.com/DataDog/dd-trace-py/pull/3966, and will be done in a future testing PR(which will essentially validate E2E) so it seemed unnecessary to go much further.

## Checklist
- [x] Title must conform to [conventional commit](https://github.com/conventional-changelog/commitlint/tree/master/%40commitlint/config-conventional).
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] Ensure tests are passing for affected code.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] PR cannot be broken up into smaller PRs.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added for fixes and features, or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.
- [ ] Add to milestone.

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -186,6 +186,13 @@ below:
        **Example:** ``DD_TRACE_SAMPLING_RULES='[{"sample_rate":0.5,"service":"my-service"}]'``
        **Note** that the JSON object must be included in single quotes (') to avoid problems with escaping of the double quote (") character.
 
+   * - ``DD_SPAN_SAMPLING_RULES``
```
Comment (ZStriker19) [inline docs/configuration.rst:189]: Should this go in a different PR?

**Comment by ZStriker19** [comment]
I pulled in changes from the single span processor PR: https://github.com/DataDog/dd-trace-py/pull/3966/files so please ignore `trace.py` changes and `test_processors.py`. Once 3966 is merged this should be mergeable.

File: ddtrace/internal/sampling.py [hunk 2]
```diff
@@ -166,3 +175,67 @@ def apply_span_sampling_tags(self, span):
...
+    def __init__(self):
+        self._rules = self._make_rules()
+
+    def get_single_span_sampling_rules(self):
+        # type: () -> List[SpanSamplingRule]
+        return self._rules
+
+    def _make_rules(self):
+        # type: () -> List[SpanSamplingRule]
+        json_rules = os.getenv("DD_SPAN_SAMPLING_RULES")
```
Comment (P403n1x87) [inline ddtrace/internal/sampling.py:192]:
```suggestion
        json_rules = os.getenv("DD_TRACER_SPAN_SAMPLING_RULES")
```
?

Comment (ZStriker19) [reply ddtrace/internal/sampling.py:192]: No, the name is correct. I also started writing that name before reviewing it though.

Comment (P403n1x87) [reply ddtrace/internal/sampling.py:192]: I wonder if this shouldn't be changed though 🤔 . In light of the effort of creating shared configuration, where would this variable stand? I'd say this is a tracer-related setting, so to me it would make sense to add the `DD_TRACER` prefix. WDYT?

Comment (ZStriker19) [reply ddtrace/internal/sampling.py:192]: I'm not strongly opinionated, but I know other tracers have already finished work using `DD_SPAN_SAMPLING_RULES`. I'll ask the RFC owner and see what they think.

Comment (ZStriker19) [reply ddtrace/internal/sampling.py:206]: It seems the RFC owner would like to keep it DD_SPAN_SAMPLING_RULES

File: ddtrace/internal/sampling.py [hunk 3]
```diff
@@ -180,3 +189,67 @@ def apply_span_sampling_tags(self, span):
         span.set_metric(_SINGLE_SPAN_SAMPLING_RATE, self._sample_rate)
         if self._max_per_second:
             span.set_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC, self._max_per_second)
+
+
+class SpanSamplingRules:
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:194]: Why have this as a class?

Based on how it is used, an instance of this class isn't really useful, it seems we only really need the behavior provided by `_make_rules`?

A few options:

- Keep this as a class for organizational reason, but move the methods to be static methods, so you so `SpanSamplingRules.get_single_span_sampling_rules()` instead
- Make it be a method level function instead of introducing the class
- Make a sub-module for span sampling to isolate the helper functions there

File: ddtrace/tracer.py [hunk 4]
```diff
@@ -205,6 +213,9 @@ def __init__(
                 sync_mode=self._use_sync_mode(),
                 headers={"Datadog-Client-Computed-Stats": "yes"} if self._compute_stats else {},
             )
+        self._single_span_sampling_rules = (
+            SpanSamplingRules().get_single_span_sampling_rules()
```
Comment (brettlangdon) [inline ddtrace/tracer.py:217]: this is the part where creating an instance of `SpanSamplingRules` feels weird, since we never actually need it to be an instance, we just need a helper to parse the env and give us a list of rules.

File: ddtrace/internal/sampling.py [hunk 5]
```diff
@@ -180,3 +189,67 @@ def apply_span_sampling_tags(self, span):
...
+
+    def _parse_rules(self, rules):
+        sampling_rules = []
+        if rules is not None:
+            json_rules = []
+            try:
+                json_rules = json.loads(rules)
+            except JSONDecodeError:
+                raise ValueError("Unable to parse DD_SPAN_SAMPLING_RULES={}".format(rules))
+            for rule in json_rules:
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:220]: we might want to assert that `json_rules` is actually a list, it could be anything after `json.loads()`

Comment (brettlangdon) [reply ddtrace/internal/sampling.py:220]: same, we might want to assert that `rule` is a dict.

(when I say "assert", I don't mean, `assert`, but a type check and then raise a `TypeError` ... ?)

File: ddtrace/internal/sampling.py [hunk 6]
```diff
@@ -180,3 +189,67 @@ def apply_span_sampling_tags(self, span):
...
+                sample_rate = rule.get("sample_rate", 1.0)
+                service = rule.get("service")
+                name = rule.get("name")
+                max_per_second = rule.get("max_per_second")
+                if service is None and name is None:
+                    raise SingleSpanSamplingError("Neither service or name specified for single span sampling rule")
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:226]: we should consider including `rule` in the message, so they know which one to look at

File: ddtrace/internal/sampling.py [hunk 7]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
         span.set_metric(_SINGLE_SPAN_SAMPLING_RATE, self._sample_rate)
         if self._max_per_second:
             span.set_metric(_SINGLE_SPAN_SAMPLING_MAX_PER_SEC, self._max_per_second)
+
+
+def make_span_sampling_rules():
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:194]:
```suggestion
def get_span_sampling_rules():
```

not sure I have a strong opinion over which verb is used in the name

File: ddtrace/internal/sampling.py [hunk 8]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+    # type: () -> List[SpanSamplingRule]
+    json_rules = os.getenv("DD_SPAN_SAMPLING_RULES")
+    if not json_rules:
+        return []
+    else:
+        return _parse_rules(json_rules)
+
+
+def _parse_rules(rules):
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:203]:
```suggestion

```

File: ddtrace/internal/sampling.py [hunk 9]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+        for rule in json_rules:
+            if rule is not dict:
+                raise TypeError("rule specified via DD_SPAN_SAMPLING_RULES is not a dictionary:{}".format(rule))
+            sample_rate = rule.get("sample_rate", 1.0)
+            service = rule.get("service")
+            name = rule.get("name")
+            max_per_second = rule.get("max_per_second")
+            if service is None and name is None:
+                raise SingleSpanSamplingError(
+                    "Neither service or name specified for single span sampling rule:{}".format(rule)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:222]:
```suggestion
                    "Neither service or name specified for single span sampling rule:{}, at least one of these must be specified".format(rule)
```

File: ddtrace/internal/sampling.py [hunk 10]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+                )
+            if (service and _unsupported_pattern(service)) or (name and _unsupported_pattern(name)):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:224]:
```suggestion
            if (service and _unsupported_pattern(service)) or (name and _unsupported_pattern(name)):
```

can we split these into two checks and raise appropriate errors to help a user if these are raised

File: ddtrace/internal/sampling.py [hunk 11]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+                )
+            except ValueError as e:
+                raise ValueError("Error creating single span sampling rule {}: {}".format(json.dumps(rule), e))
+            sampling_rules.append(sampling_rule)
+    return sampling_rules
+
+
+def _unsupported_pattern(string):
+    # We don't support pattern bracket expansion or escape character
+    unsupported_chars = ["[", "]", "\\"]
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:239]:
```suggestion
    unsupported_chars = {"[", "]", "\\"}
```

File: ddtrace/internal/sampling.py [hunk 12]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+    for char in string:
+        if char in unsupported_chars:
+            return True
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:242]: can we `raise` here instead so the user can know which character was problematic

File: ddtrace/internal/sampling.py [hunk 13]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+        try:
+            json_rules = json.loads(rules)
+            if json_rules is not list:
+                raise TypeError("DD_SPAN_SAMPLING_RULES is not list", json_rules)
+        except JSONDecodeError:
+            raise ValueError("Unable to parse DD_SPAN_SAMPLING_RULES={}".format(rules))
+        for rule in json_rules:
+            if rule is not dict:
+                raise TypeError("rule specified via DD_SPAN_SAMPLING_RULES is not a dictionary:{}".format(rule))
+            sample_rate = rule.get("sample_rate", 1.0)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:216]:
```suggestion
            sample_rate = float(rule.get("sample_rate", 1.0))
```

File: ddtrace/internal/sampling.py [hunk 14]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+            if service is None and name is None:
+                raise SingleSpanSamplingError(
+                    "Neither service or name specified for single span sampling rule:{}".format(rule)
+                )
+            if (service and _unsupported_pattern(service)) or (name and _unsupported_pattern(name)):
+                raise UnsupportedGlobPatternError
+
+            try:
+                sampling_rule = SpanSamplingRule(
+                    sample_rate=sample_rate, service=service, name=name, max_per_second=max_per_second
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:229]: `sample_rate` here is `str|float` when it's typed in SpanSamplingRule to be a `float`

also `SpanSamplingRule` should accept only float for `sample_rate` (need to update this in a separate PR)

Comment (Kyle-Verhoog) [reply ddtrace/internal/sampling.py:229]: similarly for the other types

File: ddtrace/internal/sampling.py [hunk 15]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+                )
+            except ValueError as e:
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:231]:
```suggestion
            except Exception as e:
```

File: ddtrace/internal/sampling.py [hunk 16]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+        if char in unsupported_chars:
+            return True
+    return False
+
+
+class SingleSpanSamplingError(Exception):
+    """Raised when neither a name or service are specified for a single span sampling rule"""
+
+
+class UnsupportedGlobPatternError(Exception):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:250]: Maybe worth keeping these all as `ValueError`s since they are shown to the user. We are not calling this function programmatically so there is no need to raise different classes of exception.

**Review by Kyle-Verhoog — COMMENTED** [review]
high level looks good, just a bunch of user-centric nits

File: ddtrace/internal/sampling.py [hunk 17]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+def get_span_sampling_rules():
+    # type: () -> List[SpanSamplingRule]
+    json_rules = os.getenv("DD_SPAN_SAMPLING_RULES")
+    if json_rules is None:
+        return []
+    else:
+        return _parse_rules(json_rules)
+
+
+def _parse_rules(rules):
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:203]: type hinting is needed here

File: ddtrace/internal/sampling.py [hunk 18]
```diff
@@ -180,3 +189,65 @@ def apply_span_sampling_tags(self, span):
...
+    sampling_rules = []
+    if rules is None:
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:205]: we check for this in `get_span_sampling_rules` so it should always be a string?

File: ddtrace/internal/sampling.py [hunk 19]
```diff
@@ -180,3 +192,55 @@ def apply_span_sampling_tags(self, span):
...
+                raise TypeError("DD_SPAN_SAMPLING_RULES is not list, got %r" % json_rules)
+        except JSONDecodeError:
+            raise ValueError("Unable to parse DD_SPAN_SAMPLING_RULES=%r" % json_rules_raw)
+        for rule in json_rules:
+            if not isinstance(rule, dict):
+                raise TypeError("rule specified via DD_SPAN_SAMPLING_RULES is not a dictionary:%r" % rule)
+            sample_rate = float(rule.get("sample_rate", 1.0))
+            service = rule.get("service")
+            name = rule.get("name")
+            max_per_second = int(rule.get("max_per_second")) if rule.get("max_per_second") else None
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:216]:
```suggestion
            max_per_second = int(rule.get("max_per_second", -1))
```

File: ddtrace/internal/sampling.py [hunk 20]
```diff
@@ -180,3 +192,55 @@ def apply_span_sampling_tags(self, span):
...
+            sampling_rules.append(sampling_rule)
+        return sampling_rules
+
+
+def _check_unsupported_pattern(string):
+    # We don't support pattern bracket expansion or escape character
+    unsupported_chars = ["[", "]", "\\"]
+    for char in string:
+        if char in unsupported_chars:
+            raise ValueError("Unsupported Glob pattern found, character:%r is not supported" % char)
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:239]:
```suggestion
            raise ValueError("Unsupported glob pattern found, character:%r is not supported in %r" % (char, string))
```

File: ddtrace/internal/sampling.py [hunk 21]
```diff
@@ -180,3 +192,55 @@ def apply_span_sampling_tags(self, span):
...
+                sampling_rule = SpanSamplingRule(
+                    sample_rate=sample_rate, service=service, name=name, max_per_second=max_per_second
+                )
+            except Exception as e:
+                raise ValueError("Error creating single span sampling rule {}: {}".format(json.dumps(rule), e))
+            sampling_rules.append(sampling_rule)
+        return sampling_rules
+
+
+def _check_unsupported_pattern(string):
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/sampling.py:234]: type hint

File: ddtrace/internal/sampling.py (hunk 14 above)
Comment (ZStriker19) [reply ddtrace/internal/sampling.py:226]: https://github.com/DataDog/dd-trace-py/pull/4023

**Review by mabdinur — APPROVED** [review]

**Review by Kyle-Verhoog — DISMISSED** [review]
💯 🧺

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]

File: ddtrace/internal/sampling.py [hunk 22]
```diff
@@ -178,6 +193,66 @@ def apply_span_sampling_tags(self, span):
...
+if sys.version_info >= (3, 7, 0):
+    SpanSamplingRules = TypedDict(
+        "SpanSamplingRules",
+        {
+            "name": NotRequired[str],
+            "service": NotRequired[str],
+            "sample_rate": NotRequired[float],
+            "max_per_second": NotRequired[int],
+        },
+    )
```
Comment (mabdinur) [inline ddtrace/internal/sampling.py:253]:
```suggestion
        "SpanSamplingRules",
        {
            "name": str,
            "service": str,
            "sample_rate": float,
            "max_per_second": int,
        },
        total=False,
    )
```

Since all the fields are not required we can simply set total to false: https://peps.python.org/pep-0589/#totality. With this change we can avoid using `NotRequired`

File: ddtrace/internal/sampling.py [hunk 23]
```diff
@@ -1,7 +1,15 @@
...
+import os
 import re
+import sys
 from typing import Optional
 from typing import TYPE_CHECKING
 
+
+if sys.version_info >= (3, 7, 0):
+    from typing_extensions import NotRequired
+    from typing_extensions import TypedDict
```
Comment (mabdinur) [inline ddtrace/internal/sampling.py:11]:
```suggestion
try:
   from typing import TypedDict
 except ImportError:
   from typing_extensions import TypedDict
```
TypedDict was added to [typing](https://peps.python.org/pep-0589/) in python 3.8. For `sys.version_info < (3, 8)` we should import from typing_extensions

File: ddtrace/internal/sampling.py (hunk 22 above)
Comment (mabdinur) [reply ddtrace/internal/sampling.py:253]: [NotRequired](https://peps.python.org/pep-0655/#abstract) is useful when some keys are required and others are not

**Review by mabdinur — COMMENTED** [review]
Left some suggestions. Overall it LGTM

File: ddtrace/internal/sampling.py (hunk 22 above)
Comment (Kyle-Verhoog) [reply ddtrace/internal/sampling.py:253]: I think I'd prefer `NotRequired` here since it's more explicit. If a new entry is added to the dictionary then it won't implicitly be `NotRequired`. Just my opinion though.

**Review by Kyle-Verhoog — APPROVED** [review]
lgtm with @mabdinur's typeddict comment resolved

File: ddtrace/internal/sampling.py [hunk 24]
```diff
@@ -178,6 +197,66 @@ def apply_span_sampling_tags(self, span):
...
+def _check_unsupported_pattern(string):
+    # type: (str) -> None
+    # We don't support pattern bracket expansion or escape character
+    unsupported_chars = {"[", "]", "\\"}
+    for char in string:
+        if char in unsupported_chars:
+            raise ValueError("Unsupported Glob pattern found, character:%r is not supported" % char)
+
+
+if sys.version_info >= (3, 7, 0):
```
Comment (brettlangdon) [inline ddtrace/internal/sampling.py:248]: Do we need to define an alternative for when we are not on Python 3.7?

e.g.

```python
if sys.version_info >= (3, 7, 0):
    SpanSamplingRules = TypedDict(...)
else:
    SpanSamplingRules = Dict
```

As-is this seems fine for our CI since we are using Python> 3.7 for `mypy` checking, but a customer might be using a version <3.7. Will `mypy` cause issues in those cases?

Comment (brettlangdon) [reply ddtrace/internal/sampling.py:248]: This is a nit... can we move this definition towards the top of the file.

I think we've mostly been organizing as:

1. imports
2. constants and type aliases (like this)
3. class/function definitions (everything else)

File: ddtrace/internal/sampling.py (hunk 22 above)
Comment (ZStriker19) [reply ddtrace/internal/sampling.py:257]: Given Brett's comment here: https://github.com/DataDog/dd-trace-py/pull/3977#discussion_r936585840 I think it'd be better to go with using Munir's suggestion above.

**Review by brettlangdon — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]

**Comment by majorgreys** [comment]
@Mergifyio backport 1.4
